src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1286,71 +1286,469 @@ mongoc_uri_finalize_tls (mongoc_uri_t *uri, bson_error_t *error)
+                              bsonBuildError);
+            goto fail;
+         }
+         if (!mongoc_uri_set_mechanism_properties (uri, &props)) {
+            MONGOC_URI_ERROR (error,
+                              "unexpected URI credentials BSON error when attempting to default 'GSSAPI' "
+                              "authentication mechanism property 'SERVICE_NAME' to 'mongodb': %s",
+                              bsonBuildError);
+            goto fail;
+         }
+         bson_destroy (&props);
       }
    }
-   return true;
+
+   // MONGODB-AWS
+   else if (strcasecmp (mechanism, "MONGODB-AWS") == 0) {
+      // Authentication spec: username: MAY be specified (as the non-sensitive AWS access key).
+      if (!_finalize_auth_username (username, mechanism, _mongoc_uri_finalize_allowed, error)) {
+         goto fail;
+      }
+
+      // Authentication spec: source: MUST be "$external" and defaults to "$external".
+      if (!source) {
+         bsonBuildAppend (uri->credentials, kv (MONGOC_URI_AUTHSOURCE, cstr ("$external")));
+         if (bsonBuildError) {
+            MONGOC_URI_ERROR (error,
+                              "unexpected URI credentials BSON error when attempting to default 'MONGODB-AWS' "
+                              "authentication mechanism source to '$external': %s",
+                              bsonBuildError);
+            goto fail;
+         }
+      } else if (!_finalize_auth_source_external_required (source, mechanism, error)) {
+         goto fail;
+      }
+
+      // Authentication spec: password: MAY be specified (as the sensitive AWS secret key).
+      if (!_finalize_auth_password (password, mechanism, _mongoc_uri_finalize_allowed, error)) {
+         goto fail;
+      }
+
+      // mechanism_properties are allowed for MONGODB-AWS.
+      if (!_finalize_auth_aws_mechanism_properties (mechanism_properties, error)) {
+         goto fail;
+      }
+
+      // Authentication spec: if a username is provided without a password (or vice-versa), Drivers MUST raise an error.
+      if (!username != !password) {
+         MONGOC_URI_ERROR (error, "'%s' authentication mechanism requires both a username and a password", mechanism);
+         goto fail;
+      }
+
+      // Authentication spec: if *only* a session token is provided, Drivers MUST raise an error.
+      if (!username && mechanism_properties) {
+         bson_iter_t iter;
+         if (bson_iter_init_find_case (&iter, mechanism_properties, "AWS_SESSION_TOKEN")) {
+            MONGOC_URI_ERROR (error,
+                              "%s",
+                              "'MONGODB-AWS' authentication mechanism requires AWS_SESSION_TOKEN to be accompanied by "
+                              "a username and a password");
+            goto fail;
+         }
+      }
+   }
+
+   // Invalid or unsupported authentication mechanism.
+   else {
+      MONGOC_URI_ERROR (error, "Unsupported value for \"authMechanism\": \"%s\"", mechanism);
```

Comment:
Suggest listing the supported values to help a user identify a possible typo:

```suggestion
      MONGOC_URI_ERROR (error,
                        "Unsupported value for \"authMechanism\": \"%s\". Supported values: \"SCRAM-SHA-1\", "
                        "\"SCRAM-SHA-256\", \"PLAIN\", \"MONGODB-X509\", \"GSSAPI\", \"MONGODB-AWS\"",
                        mechanism);
```